### PR TITLE
Make Intel IPC stream ops generic

### PR DESCRIPTION
**Pull request by Potochi**
Operations implemented in sof/intel/intel-ipc.c should be generic and are of a good use to IMX also. So, make them generic.

This is a rebase of #2914.

### CODE REVIEW

File: sound/soc/sof/stream-ipc.c [hunk 1]
```diff
@@ -17,39 +17,37 @@
...
 };
 
-/* Mailbox-based Intel IPC implementation */
-int intel_ipc_msg_data(struct snd_sof_dev *sdev,
-		       struct snd_pcm_substream *substream,
-		       void *p, size_t sz)
+/* Generic IPC mailbox implementation */
+void sof_ipc_msg_data(struct snd_sof_dev *sdev,
+		      struct snd_pcm_substream *substream,
+		      void *p, size_t sz)
```
Comment (dbaluta) [inline sound/soc/sof/stream-ipc.c]: @Potochi 

return type of sof_ipc_msg_data should be int. See diff. And also, 

```
 sound/soc/sof/stream-ipc.c: In function ‘sof_ipc_msg_data’:
sound/soc/sof/stream-ipc.c:42:9: error: ‘return’ with a value, in function returning void [-Werror=return-type]
   42 |  return 0;
      |         ^
sound/soc/sof/stream-ipc.c:28:6: note: declared here
   28 | void sof_ipc_msg_data(struct snd_sof_dev *sdev,
```
That's because of commit dcea805219d0276916c6ef3266fa1c4d40eced55

File: sound/soc/sof/core.c [hunk 2]
```diff
@@ -310,6 +310,7 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 	/* check all mandatory ops */
 	if (!sof_ops(sdev) || !sof_ops(sdev)->probe || !sof_ops(sdev)->run ||
 	    !sof_ops(sdev)->block_read || !sof_ops(sdev)->block_write ||
+	    !sof_ops(sdev)->mailbox_read || !sof_ops(sdev)->mailbox_write ||
```
Comment (plbossart) [inline sound/soc/sof/core.c]: I don't really follow this patch @dbaluta @Potochi
Do you have an example of a platform that would use a nonstandard sof_mailbox_read/sof_mailbox_write? 

is this coming in a follow-up PR?
And what does this have to do with the PR title "Make Intel IPC stream ops generic"

File: sound/soc/sof/sof-priv.h [hunk 3]
```diff
@@ -603,17 +603,17 @@ int sof_block_read(struct snd_sof_dev *sdev, enum snd_sof_fw_blk_type blk_type,
...
-int intel_pcm_open(struct snd_sof_dev *sdev,
-		   struct snd_pcm_substream *substream);
-int intel_pcm_close(struct snd_sof_dev *sdev,
-		    struct snd_pcm_substream *substream);
+		       const struct sof_ipc_pcm_params_reply *reply);
+
+int sof_stream_pcm_open(struct snd_sof_dev *sdev,
+			struct snd_pcm_substream *substream);
+int sof_stream_pcm_close(struct snd_sof_dev *sdev,
+			 struct snd_pcm_substream *substream);
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h:616]: This part looks good and may be beneficial to our AMD friends as well. @ajitkupandey would this change make your PR simpler?

File: sound/soc/sof/core.c (hunk 2 above)
Comment (dbaluta) [reply sound/soc/sof/core.c]: This was suggested by @ujfalusi in https://github.com/thesofproject/linux/pull/2914.

image

This commit is linked with the PR title 'Make Intel IPC stream ops generic' because when moving  `sound/soc/sof/intel/intel-ipc.c ` → ` sound/soc/sof/stream-ipc.c ` we need to use 
`snd_sof_mailbox_read` instead of `snd_sof_dsp_mailbox_read`

So, to sum up the concern is that future platforms might have the mailbox allocated in a different memory area other than the traditional memory mapped IO. 

This is what I understood. @ujfalusi is that right?

Comment (plbossart) [reply sound/soc/sof/core.c]: That's fine to future-proof, the commit message does not provide any information though. You may want to reword it to that it's clear in the commit message that the patch it's not related to any hardware addition - rather making sure we are not bound to an MMIO-type memory.

Comment (dbaluta) [reply sound/soc/sof/core.c]: Agree. We will modify the commit message.

File: sound/soc/sof/stream-ipc.c (hunk 1 above)
Comment (dbaluta) [reply sound/soc/sof/stream-ipc.c]: Thanks!

File: sound/soc/sof/core.c (hunk 2 above)
Comment (Potochi) [reply sound/soc/sof/core.c]: We updated the commit message, hopefully it is clearer now.

**Review by plbossart — APPROVED** [review]

**Review by plbossart — CHANGES_REQUESTED** [review]
ARM compilation failure?

  CC [M]  sound/soc/fsl/fsl-asoc-card.o
sound/soc/sof/imx/imx8.c:388:12: error: ‘imx8_ipc_pcm_params’ defined but not used [-Werror=unused-function]
  388 | static int imx8_ipc_pcm_params(struct snd_sof_dev *sdev,
      |            ^~~~~~~~~~~~~~~~~~~
sound/soc/sof/imx/imx8.c:380:12: error: ‘imx8_ipc_msg_data’ defined but not used [-Werror=unused-function]
  380 | static int imx8_ipc_msg_data(struct snd_sof_dev *sdev,
      |            ^~~~~~~~~~~~~~~~~

**Comment by dbaluta** [comment]
@Potochi can you have a look. Why this didn't fail on our tests? Can you check the .config file?

File: sound/soc/sof/core.c (hunk 2 above)
Comment (ajitkupandey) [inline sound/soc/sof/core.c]: AMD platform is not using mmaped IO for IPC so make it an optional ops() callback and not a mandatory one.

File: sound/soc/sof/sof-priv.h (hunk 3 above)
Comment (ajitkupandey) [reply sound/soc/sof/sof-priv.h:616]: yeah in a first look this looks helpful to us also

File: sound/soc/sof/sof-priv.h [hunk 4]
```diff
@@ -155,6 +155,14 @@ struct snd_sof_dsp_ops {
...
 			   void *src, size_t size); /* mandatory */
 
+	/* Mailbox IO */
+	void (*mailbox_read)(struct snd_sof_dev *sof_dev,
+			     u32 offset, void *dest,
+			     size_t size); /* mandatory */
+	void (*mailbox_write)(struct snd_sof_dev *sof_dev,
+			      u32 offset, void *src,
+			      size_t size); /* mandatory */
+
```
Comment (dbaluta) [inline sound/soc/sof/sof-priv.h:165]: @Potochi if we make mailbox_read /  mailbox_write optional as AMD guys suggested please modify the comment here to read 'optional' instead of 'mandatory'

@Potochi Later edit: Let's wait for a bit to understand if these are mandatory or not.

File: sound/soc/sof/core.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/core.c]: Not following @ajitkupandey. This patch precisely enables you to not use mmap IO but the mechanism of your choice.

File: sound/soc/sof/sof-priv.h (hunk 4 above)
Comment (dbaluta) [reply sound/soc/sof/sof-priv.h:165]: @ajitkupandey can you point us to the AMD implementation of ipc_msg_data? We want to understand if you can use the generic functions introduced by our commit. 

We use the abstraction of the 'mailbox' to exchange various information between DSP and Application Processor. You don't need to have IO mapped memory, it can be everything.

File: sound/soc/sof/core.c (hunk 2 above)
Comment (ajitkupandey) [reply sound/soc/sof/core.c]: yeah i saw we can implement our custom mailbox_read()/write() but we aren't using mailbox read and write at the moment. If we make it mandatory we will be forced to implement such callbacks as directly sof_mailbox_read() not going to work for us so we prefer to make it an optional dsp ops () callback for now.

File: sound/soc/sof/sof-priv.h (hunk 4 above)
Comment (ajitkupandey) [reply sound/soc/sof/sof-priv.h:165]: @dbaluta  yeah we are also using a kind of mailbox mechanism but not implemented mailbox_read()/write() .. you can refer to for ipc_msg_data() implementation https://github.com/thesofproject/linux/pull/3054/commits/1f96f249e1b26b280fa9d0122392e6cf92965ded#diff-1d060dee752ac0ca2adf00c0754be0bf9b554767e4585b6c9cbafc224a94114dR152

File: sound/soc/sof/core.c (hunk 2 above)
Comment (dbaluta) [reply sound/soc/sof/core.c]: @ajitkupandey OK. I think for now we can make this ops optional.

File: sound/soc/sof/sof-priv.h (hunk 4 above)
Comment (Potochi) [reply sound/soc/sof/sof-priv.h:165]: Fixed.

File: sound/soc/sof/core.c (hunk 2 above)
Comment (ajitkupandey) [reply sound/soc/sof/core.c]: @dbaluta @plbossart Thanks.. FYI we will implement custom mailbox_read()/write()  for AMD platform to make our code more streamline with SOF core but for now keeping these ops optional is really helpful.

**Review by dbaluta — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]
@Potochi , @dbaluta, thank you, looks nice.
